url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/decomposition/eig/watched/WatchedDoubleStepQREigen.java

```diff
@@ -0,0 +1,602 @@
+        double z22 = A.get(x2, x2);
+
+        double a11 = A.get(x1,x1);
+        double a21 = A.get(x1+1,x1);
+        double a12 = A.get(x1,x1+1);
+        double a22 = A.get(x1+1,x1+1);
+        double a32 = A.get(x1+2,x1+1);
+
+        if( normalize ) {
+            temp[0] = a11;temp[1] = a21;temp[2] = a12;temp[3] = a22;temp[4] = a32;
+            temp[5] = z11;temp[6] = z22;temp[7] = z12;temp[8] = z21;
+
+            double max = Math.abs(temp[0]);
+            for( int j = 1; j < temp.length; j++ ) {
+                if( Math.abs(temp[j]) > max )
+                    max = Math.abs(temp[j]);
+            }
+            a11 /= max;a21 /= max;a12 /= max;a22 /= max;a32 /= max;
+            z11 /= max;z22 /= max;z12 /= max;z21 /= max;
+        }
+
+        // these equations are derived when the eigenvalues are extracted from the lower right
+        // 2 by 2 matrix.  See page 388 of Fundamentals of Matrix Computations 2nd ed for details.
+        double b11,b21,b31;
+        if( useStandardEq ) {
+            b11 = ((a11- z11)*(a11- z22)- z21 * z12)/a21 + a12;
+            b21 = a11 + a22 - z11 - z22;
+            b31 = a32;
+        } else {
+            // this is different from the version in the book and seems in my testing to be more resilient to
+            // over flow issues
+            b11 = ((a11- z11)*(a11- z22)- z21 * z12) + a12*a21;
+            b21 = (a11 + a22 - z11 - z22)*a21;
+            b31 = a32*a21;
+        }
+
+        performImplicitDoubleStep(x1, x2, b11 , b21 , b31 );
+    }
+
+    /**
+     * Performs an implicit double step given the set of two imaginary eigenvalues provided.
+     * Since one eigenvalue is the complex conjugate of the other only one set of real and imaginary
+     * numbers is needed.
+     *
+     * @param x1 upper index of submatrix.
+     * @param x2 lower index of submatrix.
+     * @param real Real component of each of the eigenvalues.
+     * @param img Imaginary component of one of the eigenvalues.
+     */
+    public void performImplicitDoubleStep(int x1, int x2 , double real , double img  ) {
+        double a11 = A.get(x1,x1);
+        double a21 = A.get(x1+1,x1);
+        double a12 = A.get(x1,x1+1);
+        double a22 = A.get(x1+1,x1+1);
+        double a32 = A.get(x1+2,x1+1);
+
+        double p_plus_t = 2.0*real;
+        double p_times_t = real*real + img*img;
+
+        double b11,b21,b31;
+        if( useStandardEq ) {
+            b11 = (a11*a11 - p_plus_t*a11+p_times_t)/a21 + a12;
+            b21 = a11+a22-p_plus_t;
+            b31 = a32;
+        } else {
+            // this is different from the version in the book and seems in my testing to be more resilient to
+            // over flow issues
+            b11 = (a11*a11 - p_plus_t*a11+p_times_t) + a12*a21;
+            b21 = (a11+a22-p_plus_t)*a21;
+            b31 = a32*a21;
+        }
+
+        performImplicitDoubleStep(x1, x2, b11, b21, b31);
+    }
+
+    private void performImplicitDoubleStep(int x1, int x2,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 286-325 into a separate method.